### PR TITLE
Database.Persist.Sql: Add function transactionBegin

### PR DESCRIPTION
**Pull request by erikd**
This is a fix for the issue #949. I have built my code against this fix and it seems to do the right thing. The `withTransaction` function is less that great though. Happy to get suggestions for improvements.

Before submitting your PR, check that you've:

- [x] Bumped the version number
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

### CODE REVIEW

File: persistent/Database/Persist/Sql.hs [hunk 1]
```diff
@@ -39,6 +42,24 @@ import Database.Persist.Sql.Orphan.PersistQuery
 import Database.Persist.Sql.Orphan.PersistStore
 import Database.Persist.Sql.Orphan.PersistUnique ()
 
+-- | Begin an SQL transaction.
+--
+-- @since 2.10.2
+transactionBegin :: MonadIO m => ReaderT SqlBackend m ()
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql.hs]: This needs to be generalized to `(MonadIO m, BackendCompatible SqlBackend backend) => ReaderT backend m ()`.

File: persistent/Database/Persist/Sql.hs [hunk 2]
```diff
@@ -39,6 +42,24 @@ import Database.Persist.Sql.Orphan.PersistQuery
...
+transactionBegin = do
+    conn <- ask
+    let getter = getStmtConn conn
+    liftIO $ connBegin conn getter Nothing
+
+-- | Begin an SQL transaction.
+--
+-- @since 2.10.2
+transactionCommit :: MonadIO m => ReaderT SqlBackend m ()
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql.hs]: Ditto

File: persistent/Database/Persist/Sql.hs [hunk 3]
```diff
@@ -74,3 +95,17 @@ transactionUndoWithIsolation isolation = do
...
+--
+-- @since 2.10.2
+withTransaction :: MonadIO m => ReaderT SqlBackend m a -> ReaderT SqlBackend m a
+withTransaction action = do
+    conn <- ask
+    let getter = getStmtConn conn
+    liftIO $ connBegin conn getter Nothing
+    a <- action
+    liftIO $ connCommit conn getter
+    pure a
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql.hs]: So this isn't exception-safe. We'll get a dangling transaction if `action` throws. The function `runSqlConn` does this right:

```haskell
runSqlConn :: (MonadUnliftIO m, BackendCompatible SqlBackend backend) => ReaderT backend m a -> backend -> m a
runSqlConn r conn = withRunInIO $ \runInIO -> mask $ \restore -> do
    let conn' = projectBackend conn
        getter = getStmtConn conn'
    restore $ connBegin conn' getter Nothing
    x <- onException
            (restore $ runInIO $ runReaderT r conn)
            (restore $ connRollback conn' getter)
    restore $ connCommit conn' getter
    return x
```

But, if you're using this function (which you should be), then you're always going to be in a transaction. So doing `BEGIN` here would make a nested transaction, which is not really well supported in many databases, and likely not what is desired.

**Comment by erikd** [comment]
Thanks for the comments @parsonsmatt .

I am currently using `withPostgresqlConn` from the `persistent-postgresql`. Maybe that needs a comment to point people to `runSqlConn`. Comments on the existing `transactionSave` and `transactionUndo` functions.

**Comment by erikd** [comment]
Ah, so using `withPostgesqlConn` I should use `runSqlConn` instead of `runReaderT`!!

**Comment by erikd** [comment]
Will close this and submit a new one that just add the much needed comments.

**Comment by erikd** [comment]
> However, it's becoming clear to me that a quick type-search is going to get you from `SqlPersistT m a ~ ReaderT SqlBackend m a`, and you'll reach for `flip runReaderT sqlBackend` to do this. That must be what's happening, because I've seen this pattern at my current employer and now you're asking for it.

Yes, that is how I go there. I am now using `runSqlConn` instead of `runReaderT`. Comments submitted as #951 .

**Pull request closed without merging**